File: MyLeetCode/No543.diameter-of-binary-tree.cpp

#include<iostream>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        res = 0;
        maxDeep(root);
        return res - 1;
    }

private:
    int res = INT_MIN;
    int maxDeep(TreeNode* root) {
        if (!root) {
            return 0;
        }

        int leftMax = maxDeep(root->left);
        int rightMax = maxDeep(root->right);

        int val = leftMax + rightMax + 1;
        if (val > res) {
            res = val;
        }

        return max(leftMax, rightMax) + 1;
    }
};

int main(void) {
    Solution solution;
    TreeNode* tree1 = new TreeNode(1);
    tree1->left = new TreeNode(2);
    tree1->right = new TreeNode(3);
    tree1->left->left = new TreeNode(4);
    tree1->left->right = new TreeNode(5);


    cout << "diameter of binary of tree1 is: " << solution.diameterOfBinaryTree(tree1) << endl;

    return 0;
}

File: MyGolang/src/LeetCode/No543.diameter-of-binary-tree.go
/*
No543.diameter-of-binary-tree.go
*/
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	var ret int = 0

	var maxDep func(*TreeNode) int
	maxDep = func(node *TreeNode) int {
		if node == nil {
			return 0
		}

		leftMaxDep := maxDep(node.Left)
		rightMaxDep := maxDep(node.Right)

		// cmp max diameter
		tmpDim := leftMaxDep + rightMaxDep
		if tmpDim > ret {
			ret = tmpDim
		}

		return max(leftMaxDep, rightMaxDep) + 1
	}

	maxDep(root)
	return ret

}

func max(x, y int) int {
	if y > x {
		return y
	}
	return x
}